### PR TITLE
Color map depending on point period

### LINKED ISSUE #7: Change color map: color depending on the period
I want to generate an image like this:

image

Where the colour of the pixel depends on the point period.

@jeremy-rifkin has already implemented it in C++, so I can try to port his code to Rust.

### PR DESCRIPTION
**Pull request by josecelano**
Thanks to @jeremy-rifkin for pointing me at the solution.

TODO:

- [x] Port basic functions to calculate the period from this [C++ implementation](https://github.com/jeremy-rifkin/mandelbrot-orbits) by @jeremy-rifkin
- [x] Replace the pixel colour (greyscale) with a new colour for each period.

### CODE REVIEW

**Comment by josecelano** [comment]
I've generated an image using the function to calculate the period, but I get this image:

mandelbrot_500x500

I'm using four different colors (greyscale) for periods 1..3. The smaller the period the darker the pixel.

```rust
let color = match period {
    1 => 0,
    2 => 50,
    3 => 100,
    _ => 150,
};
```

It's similar to what I was getting in the [C example](https://github.com/josecelano/c-mandelbrot-arbitrary-precision/issues/2). Actually, a little bit worse.

**Comment by josecelano** [comment]
I think I've fixed a bug in the lambda function:

```rust
fn lambda(z: Complex<f64>,c: Complex<f64>, n: usize) -> Complex<f64> {
    let mut result = phi_prime(c);

    for iter in 1..n {
        result = result * phi_prime(phi_n(z, c, iter));
    }
    result
}
```

Line `let mut result = phi_prime(z);` replace by `let mut result = phi_prime(c);`

I think it's a potential bug I imported from [the project I copied the original implementation](https://github.com/jeremy-rifkin/mandelbrot-orbits/issues/1).

I have also removed the gradient for repelling points. It makes the image uglier but at this stage it's easier to cath bugs because I only plot "pure" set points. I can implement a nicer colormap when it works fine. If can make it work.

This is the current result:

mandelbrot_5000x5000

**Comment by josecelano** [comment]
It seems for some points we know they belong to the set, but we can't calculate the period. I've changed the colormap to plot those points in grey instead of white. And the result is significatively better.

mandelbrot_4096x4096

**Comment by jeremy-rifkin** [comment]
> I've generated an image using the function to calculate the period, but I get this image:

Oh interesting! 😄 I'm excited to see this being implemented.

These are cool images you're producing, they look somewhat similar to the buggy Images I produced while trying to make an algorithm to do this. The real treasure is the bugs we make along the way.

It's entirely possible my implementation has bugs, and I think there may have been an inefficiency I never resolved.

**Comment by jeremy-rifkin** [comment]
Oh, try increasing the iteration count. If I remember correctly I had to use a rather high iteration count for the escape time calculation to get correct detail throughout the entirety of each of the bulbs. At the moment each bulb is correctly colored at its very center.

**Comment by josecelano** [comment]
> Oh, try increasing the iteration count. If I remember correctly I had to use a rather high iteration count for the escape time calculation to get correct detail throughout the entirety of each of the bulbs. At the moment each bulb is correctly colored at its very center.

Yes, I've already tried that. The thing is the point is well detected in the escape_time function but the period it's not calculated well. In fact, the lightest grey near the limit are the points that belong to the Set (escape_time function works well) but the `calculate_period` function does not work. So the problem must be in the `lambda` function.

I've been:

- Reading some parts of this [paper](https://www.acsu.buffalo.edu/~adamcunn/downloads/MandelbrotSet.pdf) you mentioned.
- And playing with [your amazing project to plot trajectories](https://rifkin.dev/projects/mandelbrot-trajectories/).
- And also checking how the orbits functions looks like https://mandelbrot-set-periods.online/

And I'm reaching this conclusion:

- The further the point is from the bulb center the longer it takes to the trajectory to reach the final period.
- For some reason the `lambda` function is detecting the first period before the period gets stable at its final value.

But I do not find a way to improve `lambda` precision since it does not depends on iterations. It's a fixed amount of iterations.

image

**Comment by josecelano** [comment]
I could fix the problem for period 2 using a hardcoded formula in lambda function:

image


Strangely it works for period 2 but not for period 1:

```rust
fn lambda(z: Complex<f64>,c: Complex<f64>, n: usize) -> Complex<f64> {

    if n == 1 {
        //  2z
        return Complex { re: 2.0, im: 0. } * c;
    }

    if n == 2 {
        //  4(c + 1) 
        return Complex { re: 4.0, im: 0. } * (c + 1.);
    }

    // n > 2

    let mut result = phi_prime(c);
    for iter in 1..n {
        result = result * phi_prime(phi_n(z, c, iter));
    }
    result
}
```

I've also change the color map to make a clear difference between period 1 and 2.

**Comment by josecelano** [comment]
I do not know why the previous patch worked. It's not based on the lambda but directly in the function to detect if the point has period 1 or 2. I have removed it because I want to use the multiplier.

**Comment by jeremy-rifkin** [comment]
> The thing is the point is well detected in the escape_time function but the period it's not calculated well.

The iteration count is more important than just detecting whether a point is inside or outside of the mandelbrot set, and now I'm seeing the error here!

For period 1 or period 2 orbits we can hardcode a simple polynomial check as you've done above. Other bulbs with higher periods would require higher order polynomials which aren't practical to hardcode or compute on the fly (well, maybe there's some clever thing that could be done here).

`lambda` is an analytical function computing a derivative of a high-order polynomial at a point and it's key that the input point lies on an orbit.

It can often take quite some time for iterations of $z_{n+1} = z_n^2 + c$ to converge to points close enough to their orbits to be used in `lambda`. I have a tool here https://rifkin.dev/projects/mandelbrot-trajectories/ to view these orbits. You'll see with two example `c` values (the red points) which are near the edges of period 1 and period 3 bulbs that it takes time to converge to either a single point or a period 3 orbit respectively:

image

image

So TL;DR: The iterations are important and as noted in the other issue it is important that the final value of `z` from thousands of iterations of the escape time function is used for lambda, as opposed to the initial value `c`.

I will try to make this more clear in the README of my project, the importance of the final `z` value is somewhat subtle.

**Comment by josecelano** [comment]
hi @jeremy-rifkin thank you very much for your help!

I was totally wrong. I thougth you should use the "c" point in the "lambda" function instead of the nth iteration of z²+c.

And the second error was the [issue](https://github.com/jeremy-rifkin/mandelbrot-orbits/issues/1) I opened on your repo. I did not realize [here](https://github.com/jeremy-rifkin/mandelbrot-orbits/blob/master/src/main.cpp#L138-L142) that you were using the final "z" value.

After fixing that, It works!!!

mandelbrot_2048x2048

Before merging this PR, I will clean the code and explain the formulas in the README.

I will also add another issue to allow RGB colours in the future.

**Comment by josecelano** [comment]
I have added a question in [math.stackexchange.com](https://math.stackexchange.com/q/4502546/1082376). It would be nice to find a more formal mathematical definition and explanation. And maybe easier to follow for a non-mathematitian.

**Comment by josecelano** [comment]
@jeremy-rifkin If case you are going to extend your README I have another question. I do not know why this loop limit is `max_period`:

```c++
bool is_period(const int n, std::complex<fp> z, const std::complex<fp> c) {
	for(int i = 0; i < max_period; i++) {
		if(std::abs(lambda(n, z, c)) >= 1) {
			return false;
		}
		z = z * z + c;
	}
	return true;
}
```

I do not see that in the formulas.

```c++
point_descriptor mandelbrot(fp x, fp y) {
        // ...

	for(int p = 1; p <= max_period; p++) {
		if(is_period(p, z, c)) {
			return {false, 0, p};
		}
	}
```

For me, it would be very helpful to know which exactly math formula is implementing each function. I can't map your implementation to the professor's slides.

**Comment by josecelano** [comment]
The beast at 16Kx16K pixels.

mandelbrot_16384x16384

**Comment by josecelano** [comment]
Hi @jeremy-rifkin, I tried to summarize all the formulas used in your solution.

I do not know if I'm using the right mathematical notation, but I tried to use different names for variables that could represent different concepts. For example, "z" is de complex "0", and also a parameter for the function z²+c and also the result, and even also represents a value after iterating n times z²+c which can be used with the lambda function.

mandelbrot-periods-with-lambda

**Comment by jeremy-rifkin** [comment]
@josecelano This looks great to me!
